### PR TITLE
Path.Kind cleanup

### PR DESCRIPTION
**Pull request by aalekseyev**
Clean up various things that were matching on path kind.
Rewrite the two things that were actually using `Path.kind` or `Path.kind_raw` to match on path directly.
Rename Path.Kind -> Path.Build.Kind because it's only used to set the build directory root now.

And rename its constructor Local -> In_source_dir to remove the confusion between the Local module and this thing.

### CODE REVIEW

**Comment by aalekseyev** [comment]
This cleanup didn't work out as originally planned: I hoped to merge all the 4 places where we currently duplicate the logic of `Utils.analyse_target`. Sadly, eliminating even one of them drove me to c5503869ecff9ee893e3996c463fd954be61e7fd, which seems like a diff too big (and too green). Let's settle for a smaller change for now.

File: src/stdune/path.mli [hunk 1]
```diff
@@ -97,28 +89,44 @@ module Build : sig
...
   val drop_build_context_exn : t -> Source.t
 
+  (** [Source.t] here is a lie in some cases: consider when the context name
+      happens to be ["install"] or [".alias"]. *)
   val extract_build_context  : t -> (string * Source.t) option
   val extract_build_context_exn  : t -> (string * Source.t)
   val extract_build_context_dir : t -> (t * Source.t) option
   val extract_build_context_dir_exn : t -> (t * Source.t)
 
+  (** This function does the same as [extract_build_context], but gas a
```
Comment (rgrinberg) [inline src/stdune/path.mli]: s/gas/has/

File: src/stdune/path.ml [hunk 2]
```diff
@@ -686,8 +676,8 @@ module Build = struct
         ]
 
   let is_alias_stamp_file s =
-    String.is_prefix (Local.to_string s) ~prefix:".aliases/"
-
+    String.is_prefix (Local.to_string s) ~prefix:".aliases/"
```
Comment (rgrinberg) [inline src/stdune/path.ml]: Unrelated, but I think we should just move this function the alias module.

Comment (aalekseyev) [reply src/stdune/path.ml]: Added a CR-someday comment.

File: src/stdune/path.mli (hunk 1 above)
Comment (aalekseyev) [reply src/stdune/path.mli]: Thanks. Fixed.

**Pull request merged**